source/server/guarddog_impl.h

```diff
@@ -0,0 +1,78 @@
+#pragma once
+
+#include "envoy/server/configuration.h"
+#include "envoy/server/guarddog.h"
+#include "envoy/server/watchdog.h"
+#include "envoy/stats/stats.h"
+
+#include "common/common/logger.h"
+#include "common/common/thread.h"
+#include "common/event/libevent.h"
+
+namespace Server {
+/**
+ * This feature performs deadlock detection stats collection & enforcement.
+ *
+ * It launches a thread that scans at an interval the minimum of the configured
+ * intervals.  If it finds starved threads or suspected deadlocks it will take
+ * the appropriate action depending on the config parameters described below.
+ *
+ * Thread lifetime is tied to GuardDog object lifetime (RAII style).
+ */
+class GuardDogImpl : public GuardDog {
+public:
+  /**
+   * @param stats_store Statistics store to write watchdog_miss and
+   * watchdog_mega_miss events into.
+   * @param config Configuration object.
+   *
+   * See the configuration documentation for details on the timeout settings.
+   */
+  GuardDogImpl(Stats::Store& stats_store, const Server::Configuration::Main& config,
```

Comment:
I think you mean Stats::Scope& which looks like an updated version of Stats::Store& ?  Is there anything special that needs to be done with regards to the thread local storage mentioned here since this is incrementing the counter in a new thread?